nengo/builder/operator.py

```diff
@@ -696,7 +698,7 @@ def step_dotinc():
             mat_A = self.bsr_matrix((A, self.indices, self.indptr))
             inc = mat_A.dot(X)
             if self.reshape:
-                inc = np.asarray(inc).reshape(Y.shape)
+                inc = np.asarray(inc, dtype=rc.sim_dtype).reshape(Y.shape)
```

Comment:
`inc` should be automatically cast to `Y's` dtype in the `+=` below (if possible, and if it's not possible then casting it here won't help either).

Reply:
I thought that with binary operations, NumPy will promote lower precision arrays to higher precision so that the result doesn't lose information. But maybe it doesn't matter when doing `+=`? I'm not sure but my inclination is to use the right dtype as quickly as possible...

Reply:
I don't think there's any reason not to cast here, though I think things would've been fine without it, too.

Reply:
> I thought that with binary operations, NumPy will promote lower precision arrays to higher precision so that the result doesn't lose information.

Since this is an in-place operation, it won't change the type of the base array (which is what we want).

> I don't think there's any reason not to cast here,

Allowing numpy to implicitly cast it as part of the += op will be faster than adding an explicit cast step.